File: Java/PetShop.java
import java.util.*; // Mengimpor semua class dari package java.util

// Class PetShop untuk mengelola daftar produk di dalam toko hewan peliharaan
public class PetShop {
    // List untuk menyimpan produk dalam bentuk Map<String, Object>
    private List<Map<String, Object>> inventory = new ArrayList<>();
    private Scanner sc = new Scanner(System.in); // Scanner untuk input pengguna

    // Fungsi untuk menambahkan produk baru ke dalam inventory
    public void tambahProduk() {
        System.out.print("ID Produk: ");
        int id = sc.nextInt(); // Meminta input ID produk
        sc.nextLine(); // Menghapus newline dari buffer
        System.out.print("Nama Produk: ");
        String nama = sc.nextLine(); // Input nama produk
        System.out.print("Kategori: ");
        String kategori = sc.nextLine(); // Input kategori produk
        System.out.print("Harga: ");
        double harga = sc.nextDouble(); // Input harga produk

        // Membuat objek produk dalam bentuk Map
        Map<String, Object> produk = new HashMap<>();
        produk.put("id", id);
        produk.put("nama", nama);
        produk.put("kategori", kategori);
        produk.put("harga", harga);

        inventory.add(produk); // Menambahkan produk ke dalam list
        System.out.println("Produk berhasil ditambahkan!");
    }

    // Fungsi untuk menampilkan semua produk dalam inventory
    public void tampilkanProduk() {
        if (inventory.isEmpty()) { // Mengecek apakah inventory kosong
            System.out.println("Tidak ada produk dalam inventory.");
        } else {
            System.out.println("\n===== DAFTAR PRODUK =====");
            for (Map<String, Object> produk : inventory) { // Iterasi setiap produk
                displayInfo(produk);
            }
        }
    }

    // Fungsi untuk mengedit produk berdasarkan ID
    public void editProduk() {
        System.out.print("Masukkan ID produk yang ingin diubah: ");
        int id = sc.nextInt();
        sc.nextLine(); // Membersihkan buffer input
        int i = 0;
        boolean found = false;

        while (i < inventory.size() && !found) { // Iterasi sampai produk ditemukan
            if ((int) inventory.get(i).get("id") == id) {
                System.out.print("Nama Baru: ");
                inventory.get(i).put("nama", sc.nextLine());
                System.out.print("Kategori Baru: ");
                inventory.get(i).put("kategori", sc.nextLine());
                System.out.print("Harga Baru: ");
                inventory.get(i).put("harga", sc.nextDouble());
                found = true;
                System.out.println("Produk berhasil diperbarui!");
            }
            i++;
        }

        if (!found) {
            System.out.println("Produk dengan ID " + id + " tidak ditemukan!");
        }
    }

    // Fungsi untuk menghapus produk berdasarkan ID
    public void hapusProduk() {
        System.out.print("Masukkan ID produk yang ingin dihapus: ");
        int id = sc.nextInt();
        int i = 0;
        boolean found = false;

        while (i < inventory.size()) {
            if ((int) inventory.get(i).get("id") == id) {
                inventory.remove(i); // Menghapus produk dari list
                System.out.println("Produk berhasil dihapus!");
                found = true;
                break;
            }
            i++;
        }

        if (!found) {
            System.out.println("Produk dengan ID " + id + " tidak ditemukan!");
        }
    }

    // Fungsi untuk mencari produk berdasarkan nama
    public void cariProduk() {
        System.out.print("Masukkan Nama Produk yang dicari: ");
        sc.nextLine(); // Menghapus newline dari buffer
        String nama = sc.nextLine().toLowerCase();
        int i = 0;
        boolean found = false;

        while (i < inventory.size() && !found) {
            if (((String) inventory.get(i).get("nama")).toLowerCase().equals(nama)) {
                System.out.println("Produk ditemukan:");
                displayInfo(inventory.get(i));
                found = true;
            }
            i++;
        }

        if (!found) {
            System.out.println("Produk dengan nama '" + nama + "' tidak ditemukan!");
        }
    }

    // Fungsi untuk menampilkan informasi produk
    private void displayInfo(Map<String, Object> produk) {
        System.out.println("ID: " + produk.get("id"));
        System.out.println("Nama Produk: " + produk.get("nama"));
        System.out.println("Kategori: " + produk.get("kategori"));
        System.out.printf("Harga: Rp %.2f\n", (double) produk.get("harga"));
        System.out.println("--------------------------");
    }
}

File: Python/PetShop.py
# Class untuk mengelola data produk di dalam PetShop
class PetShop:
    def __init__(self):
        """
        Konstruktor untuk inisialisasi daftar produk dalam bentuk list.
        Setiap produk akan disimpan dalam dictionary.
        """
        self.inventory = []

    def tambah_produk(self):
        """
        Fungsi untuk menambahkan produk baru ke dalam inventory.
        """
        # Meminta input dari pengguna
        id = int(input("ID Produk: "))  # ID unik untuk produk
        nama = input("Nama Produk: ")  # Nama produk
        kategori = input("Kategori: ")  # Kategori produk
        harga = float(input("Harga: "))  # Harga produk (dikonversi ke float)

        # Menyimpan produk dalam dictionary dan menambahkannya ke inventory
        self.inventory.append({
            "id": id,
            "nama": nama,
            "kategori": kategori,
            "harga": harga
        })
        print("Produk berhasil ditambahkan!")

    def tampilkan_produk(self):
        """
        Fungsi untuk menampilkan semua produk dalam inventory.
        """
        if not self.inventory:  # Mengecek apakah inventory kosong
            print("Tidak ada produk dalam inventory.")
        else:
            print("\n===== DAFTAR PRODUK =====")
            for produk in self.inventory:  # Loop untuk menampilkan tiap produk
                self.display_info(produk)

    def edit_produk(self):
        """
        Fungsi untuk mengedit produk berdasarkan ID.
        """
        id = int(input("Masukkan ID produk yang ingin diubah: "))  # Meminta ID produk
        found = False  # Menandai apakah produk ditemukan
        i = 0  # Variabel indeks untuk iterasi

        while i < len(self.inventory) and not found:  # Loop sampai produk ditemukan atau inventory habis
            if self.inventory[i]["id"] == id:
                # Memperbarui informasi produk
                self.inventory[i]["nama"] = input("Nama Baru: ")
                self.inventory[i]["kategori"] = input("Kategori Baru: ")
                self.inventory[i]["harga"] = float(input("Harga Baru: "))
                found = True  # Produk ditemukan
                print("Produk berhasil diperbarui!")
            i += 1  # Pindah ke produk berikutnya

        if not found:  # Jika produk tidak ditemukan
            print(f"Produk dengan ID {id} tidak ditemukan!")

    def hapus_produk(self):
        """
        Fungsi untuk menghapus produk berdasarkan ID.
        """
        id = int(input("Masukkan ID produk yang ingin dihapus: "))
        i = 0  # Variabel indeks untuk iterasi

        while i < len(self.inventory):  # Loop sampai produk ditemukan
            if self.inventory[i]["id"] == id:
                del self.inventory[i]  # Menghapus produk dari list
                print("Produk berhasil dihapus!")
                return  # Keluar setelah menghapus produk
            i += 1  # Pindah ke produk berikutnya

        print(f"Produk dengan ID {id} tidak ditemukan!")  # Jika produk tidak ditemukan

    def cari_produk(self):
        """
        Fungsi untuk mencari produk berdasarkan nama.
        """
        nama = input("Masukkan Nama Produk yang dicari: ").lower()  # Input nama dalam huruf kecil
        found = False  # Menandai apakah produk ditemukan
        i = 0  # Variabel indeks

        while i < len(self.inventory) and not found:
            if self.inventory[i]["nama"].lower() == nama:
                print("Produk ditemukan:")
                self.display_info(self.inventory[i])  # Menampilkan produk
                found = True  # Produk ditemukan
            i += 1  # Pindah ke produk berikutnya

        if not found:  # Jika produk tidak ditemukan
            print(f"Produk dengan nama '{nama}' tidak ditemukan!")

    def display_info(self, produk):
        """
        Fungsi untuk menampilkan informasi produk.
        :param produk: Dictionary yang menyimpan data produk.
        """
        print(f"ID: {produk['id']}")
        print(f"Nama Produk: {produk['nama']}")
        print(f"Kategori: {produk['kategori']}")
        print(f"Harga: Rp {produk['harga']:.2f}")
        print("--------------------------")
